### PR TITLE
Use 'best_loss' corresponding to avg valid loss for early stopping

### LINKED ISSUE #509: Unexpected early stopping behaviour
I'm currently working on finding a good configuration for training with early stopping using ModernMT (which you may have seen from #508).

In doing so, I am getting somewhat confused with the current loss values reported in the logs, and the loss values used for early stopping. In one instance I train a network with the following command:
```bash
./mmt create -e DEEN_early_stop_fp16 de en data/DEEN/ \
  --arch transformer_mmt_base \
  --max-tokens 4096 \
  --seed 42 \
  --fp16 \
  --checkpoints-num 2 \
  --save-interval 1 \
  --validate-interval 1 \
  --no-epoch-checkpoints \
  --save-interval-updates 10000
```
And now here is my unexpected behaviour in the form of an excerpt from the 'training.log'.
```
...
| epoch 011 | valid on 'valid' subset | loss 3.337 | nll_loss 1.518 | ppl 2.86 | num_updates 90000 | best_loss 3.33733
| epoch 012 | valid on 'valid' subset | loss 3.321 | nll_loss 1.504 | ppl 2.84 | num_updates 100000 | best_loss 3.3213
| epoch 013 | valid on 'valid' subset | loss 3.305 | nll_loss 1.492 | ppl 2.81 | num_updates 110000 | best_loss 3.30541
| epoch 014 | valid on 'valid' subset | loss 3.298 | nll_loss 1.481 | ppl 2.79 | num_updates 120000 | best_loss 3.29849
| epoch 016 | valid on 'valid' subset | loss 3.290 | nll_loss 1.477 | ppl 2.78 | num_updates 130000 | best_loss 3.2899
2019-10-01 23:55:08,016 [INFO] - Stop criterion: current_loss = 3.671257, previous_loss = 3.655064
2019-10-01 23:55:08,017 [INFO] - Training interrupted by termination policy: validation loss has reached its plateau
```

The [criterion](https://github.com/modernmt/modernmt/blob/master/cli/train.py#L76) is using a loss window for previous/current loss, based on the `--checkpoints-num` arg, which in my case is 2. From this implementation, I still expect to see much lower values used for the criterion, and I similarly would not expect it to stop training, as I see the validation loss steadily decreasing.

Is my combination of options producing some unintended behaviour, or is validation loss from the tensorboard events somehow not aligned with the validation loss reported in the training.log?

From other experiments without early-stopping I would expect this setup to train for almost twice as long before reaching a validation plateau.

### PR DESCRIPTION
**Pull request by ecly**
I have verified this to be working as early stopping would be expected to work, by simply training a model.

The existing implementation stopped training after ~120k iterations, producing a BLEU of about 34 for my test set, whereas the new code trains for ~380k iterations, achieving 35.7 BLEU points. 

For context and intuition behind the changes, see #509 which also links to corresponding fairseq ticket regarding the 'best_loss'.

If you agree with this change, I hope it can be released pretty fast, as I would consider this a pretty crucial bug.

Closes #509

### CODE REVIEW

**Comment by nicolabertoldi** [comment]
@ecly 

thank you very much for your investigation and your PR.

We will take into account asap.

**Comment by ecly** [comment]
Any update on including this? It should be crucial for reliable early stopping.

**Comment by mzeidhassan** [comment]
Any update on this? I don't think it's merged already.

**Comment by ecly** [comment]
@mzeidhassan indeed not merged yet sadly - despite the severity of this issue on the performance of early stopping

**Comment by mzeidhassan** [comment]
Hi @ecly,

I made the change on my local version of MMT. 
Definitely, the training is taking longer, but I noticed this.

> | epoch 002:  47100 / 52284 loss=3.640, nll_loss=1.930, ppl=3.81, wps=13663, ups=1, wpb=10578.160, bsz=448.517, num_updates=99385, lr=0.000100309, gnorm=0.602, clip=0.000, oom=0.000, wall=77162, train_wall=73805
| epoch 002:  47200 / 52284 loss=3.640, nll_loss=1.930, ppl=3.81, wps=13663, ups=1, wpb=10577.820, bsz=448.490, num_updates=99485, lr=0.000100258, gnorm=0.602, clip=0.000, oom=0.000, wall=77238, train_wall=73878
| epoch 002:  47300 / 52284 loss=3.640, nll_loss=1.930, ppl=3.81, wps=13664, ups=1, wpb=10577.692, bsz=448.465, num_updates=99585, lr=0.000100208, gnorm=0.602, clip=0.000, oom=0.000, wall=77313, train_wall=73952
| epoch 002:  47400 / 52284 loss=3.640, nll_loss=1.930, ppl=3.81, wps=13664, ups=1, wpb=10577.817, bsz=448.539, num_updates=99685, lr=0.000100158, gnorm=0.602, clip=0.000, oom=0.000, wall=77390, train_wall=74027
| epoch 002:  47500 / 52284 loss=3.640, nll_loss=1.930, ppl=3.81, wps=13665, ups=1, wpb=10577.717, bsz=448.481, num_updates=99785, lr=0.000100108, gnorm=0.602, clip=0.000, oom=0.000, wall=77465, train_wall=74101
| epoch 002:  47600 / 52284 loss=3.640, nll_loss=1.930, ppl=3.81, wps=13665, ups=1, wpb=10577.721, bsz=448.541, num_updates=99885, lr=0.000100058, gnorm=0.602, clip=0.000, oom=0.000, wall=77541, train_wall=74175
| epoch 002:  47700 / 52284 loss=3.640, nll_loss=1.929, ppl=3.81, wps=13666, ups=1, wpb=10577.651, bsz=448.504, num_updates=99985, lr=0.000100008, gnorm=0.602, clip=0.000, oom=0.000, wall=77617, train_wall=74249
| epoch 002 | valid on 'valid' subset | loss 3.446 | nll_loss 1.608 | ppl 3.05 | num_updates 100000 | best_loss 3.44052
2019-11-13 04:14:50,065 [INFO] - Stop criterion: current_loss = 3.449132, previous_loss = 3.451206
| saved checkpoint /home/mzeid/00_MMT_4/modernmt/runtime/transnat_base/tmp/training/_temp_train/nn_model/checkpoint_2_100000.pt (epoch 2 @ 100000 updates) (writing took 6.918522834777832 seconds)
| epoch 002:  47800 / 52284 loss=3.640, nll_loss=1.929, ppl=3.81, wps=13663, ups=1, wpb=10577.799, bsz=448.480, num_updates=100085, lr=9.99575e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=77701, train_wall=74322
| epoch 002:  47900 / 52284 loss=3.640, nll_loss=1.929, ppl=3.81, wps=13664, ups=1, wpb=10577.669, bsz=448.564, num_updates=100185, lr=9.99076e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=77777, train_wall=74397
| epoch 002:  48000 / 52284 loss=3.639, nll_loss=1.929, ppl=3.81, wps=13665, ups=1, wpb=10577.850, bsz=448.599, num_updates=100285, lr=9.98578e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=77853, train_wall=74471
| epoch 002:  48100 / 52284 loss=3.639, nll_loss=1.929, ppl=3.81, wps=13665, ups=1, wpb=10578.175, bsz=448.693, num_updates=100385, lr=9.98081e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=77930, train_wall=74546
| epoch 002:  48200 / 52284 loss=3.639, nll_loss=1.929, ppl=3.81, wps=13666, ups=1, wpb=10577.961, bsz=448.599, num_updates=100485, lr=9.97584e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=78005, train_wall=74620
| epoch 002:  48300 / 52284 loss=3.639, nll_loss=1.929, ppl=3.81, wps=13667, ups=1, wpb=10577.992, bsz=448.545, num_updates=100585, lr=9.97088e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=78080, train_wall=74693
| epoch 002:  48400 / 52284 loss=3.639, nll_loss=1.929, ppl=3.81, wps=13667, ups=1, wpb=10578.229, bsz=448.524, num_updates=100685, lr=9.96592e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=78156, train_wall=74768
| epoch 002:  48500 / 52284 loss=3.639, nll_loss=1.929, ppl=3.81, wps=13668, ups=1, wpb=10578.280, bsz=448.484, num_updates=100785, lr=9.96098e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=78233, train_wall=74843
| epoch 002:  48600 / 52284 loss=3.639, nll_loss=1.928, ppl=3.81, wps=13668, ups=1, wpb=10578.154, bsz=448.444, num_updates=100885, lr=9.95604e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=78308, train_wall=74916
| epoch 002:  48700 / 52284 loss=3.639, nll_loss=1.928, ppl=3.81, wps=13669, ups=1, wpb=10578.137, bsz=448.509, num_updates=100985, lr=9.95111e-05, gnorm=0.602, clip=0.000, oom=0.000, wall=78384, train_wall=74990
| epoch 002 | valid on 'valid' subset | loss 3.444 | nll_loss 1.605 | ppl 3.04 | num_updates 101000 | best_loss 3.44052


Is it normal for the lr to jump from `lr=0.000100008` to `lr=9.95111e-05`? Is this a side effect from that change?

Maybe, the change didn't trigger the stopping criteria set by MMT? Any idea?

**Comment by mzeidhassan** [comment]
It's worth-mentioning thought that the loss keeps decreasing in this line:

`Stop criterion: current_loss = 3.384518, previous_loss = 3.385031`

Same applies to best_loss values as well. Screen shots attached.

best_loss
loss

**Comment by ecly** [comment]
@mzeidhassan  I would certainly still wait. Your tensorboard looks fine, and still indicates steady progress. You should not worry about the validation loss being lower than your training loss, as this is expected due to the use of dropout. Dropout is enabled for training but not for validation, so you are bound to get better validation loss scores.

**Comment by ecly** [comment]
After the fix for early stopping, my models trained for about 2x as long as they did before, but also achieved much better scores, so it is worth the wait to let them train. Hopefully the first round of training will increase your confidence in the early stopping, such that you for future runs don't even need to monitor the loss, as the system just halts as soon as it reaches convergence.

**Comment by mzeidhassan** [comment]
Thanks @ecly for your advice. I appreciate it. I am keeping it running for now.
One last thing:
I assume MMT stopped automatically after it reaches convergence without your intervention, right? Or did you need to stop it manually? 

Thanks again!

**Comment by mzeidhassan** [comment]
Actually, training has just completed. 
Indeed, it took about 2x the time the previous training took. 
I will check its quality compared to the previous engine and let you know if I have any issues.

Thanks again @ecly for your support.

**Comment by ecly** [comment]
@mzeidhassan yes indeed - the training will stop on its own! :-) Feel free to keep my updated on its performance!

**Comment by EtienneAb3d** [comment]
I think the stop criterion of MMT is not a good one. On my conditions, it often stops way too early. I use to set the `--train-steps` parameter to enforce it to do at least 10 epochs.

**Comment by ecly** [comment]
@EtienneAb3d with the fixes in this PR, the stop criterion should be a good one. Without this fix, I agree that it is a bad one. It stops far too early.

**Comment by mzeidhassan** [comment]
@EtienneAb3d, I agree with you. Thanks to @ecly for the PR. My initial training stopped very early, even before finishing the second epoch. With the fix, it reached the 5th epoch, and I see improvement in the output quality. 

I hope there is a solid formula that can tell how many training steps or epochs you should run for a specific word count for example, but I don't think there is anything like that yet. So, the stop criterion is a good way to go. Hope MMT team will look into this.

**Comment by ecly** [comment]
@mzeidhassan sadly there is no such formula, and I would say that some variation of early stopping should always be able to perform better on average. Loss based validation has some shortcomings, and it is possible that a BLEU based validation criterion could be more indicative. I have used something similar in the past for which I saw slight improvement over loss based early stopping.

Either way, I think it is fair to trust the early stopping from this PR, as being very reasonable when it comes to stopping the training. One future addition that would be nice is support for patience. It looks like `fairseq` will support both early stopping and patience in the [near future](https://github.com/pytorch/fairseq/pull/904), so it may be that we can start using their implementation instead of this somewhat hacky implementation instead.

**Comment by ecly** [comment]
@nicolabertoldi @davidecaroselli anything blocking this from being merged. If you're doing a release 4.2.1 this would be great to get in.

**Comment by davidecaroselli** [comment]
Hi @ecly 
we wanted to run some internal tests with our very large training, and everything seems good now. I'm merging this PR and it will be part of next release.

Thanks!

**Comment by ecly** [comment]
@davidecaroselli all good - I understand. Thank you very much!
